Nation spawns

MartenM opened the pull request:
Nation spawns has been requested quite a lot, and there is even a to-do item for it on the projects page.
I decided to have some fun and try implementing it my self.

This is the result. I have tested it and it seems to work but I am not sure on the backwards-com since this does involve some data changes.

**Issues known so far:**
- Might not go well with older versions of Towny due to database changes.
- Messages other then in English are not supported.

**Sidenotes:**
- This implementation is based of the /t spawn command.

**References:**
- Issue #1163

Please be kind to me. This is my first real pull request to an open source project :heart:

LlmDl commented on src/english.yml:
```diff
@@ -71,6 +71,7 @@ nation_help_6: 'Create a new nation'
 nation_help_7: 'List the king commands'
 nation_help_8: 'Create a new nation'
 nation_help_9: 'List all online residents in nation'
+nation_help_10: 'Teleport to the nations spawn.'
```
This will have to go to the end of the language file so that translators will know to translate it.

LlmDl commented on src/english.yml:
```diff
@@ -245,6 +246,7 @@ msg_peasant_right: '&bA peasant hasn''t the right to force his leaders into the
 msg_cost_spawn: '&bYou were charged %s to teleport to the town''s spawn.'
 msg_cost_spawn_refund: '&bYou were refunded the charge for using town spawn.'
 msg_town_spawn_warmup: '&bWaiting to teleport...'
+msg_nation_spawn_warmup: '&bWaiting to teleport...'
```
Also has to go to the end of the language file.

LlmDl commented on src/english.yml:
```diff
@@ -369,6 +371,7 @@ msg_cache_block_error_town_allies: '&cAllies aren''t allowed to %s.'
 msg_cache_block_error_town_outsider: '&cOutsiders aren''t allowed to %s.'
 
 msg_err_cant_afford_tp_town: '&cCannot afford to teleport to %s.'
+msg_err_cant_afford_tp_nation: '&cCannot afford to teleport to %s.'
```
All of these new language strings.

LlmDl commented on src/english.yml:
```diff
@@ -627,6 +630,8 @@ msg_err_nation_no_requests: '&bNation does not have any ally requests!'
 msg_err_nation_sent_too_many_invites: '&bNation has sent too many invites, revoke some before trying again!'
 msg_err_nation_sent_too_many_requests: '&bNation has sent too many requests, revoke some before trying again!'
 msg_err_nation_has_too_many_requests: '&bNation %s has too many requests, ask them to decline some, before requesting again.'
+msg_err_nationspawn_has_not_been_set: '&cNation spawn has not been set.'
```
These too.

MartenM commented:
Okay will make a commit on the language files.

MartenM commented on src/com/palmergames/bukkit/towny/command/NationCommand.java:
```diff
@@ -1847,4 +1858,206 @@ public void nationRename(Player player, Nation nation, String newName) {
+                }
+            }
+
+            nationSpawnPermission.checkIfAllowed(plugin, player, resident.getTown());
```
Could you also check this line?
I am not too sure if this is working as intended. Not too familiar with townies permission handling.
Maybe I should just put in a check manually my self instead of using the TownSpawnLevel for this.

LlmDl replied:
This you can leave as it is for now.

LlmDl commented:
I'll be going over the whole thing closely soon, I just wanted to point out the language string stuff so it could get fixed beforehand.

LlmDl commented on src/com/palmergames/bukkit/towny/command/NationCommand.java:
```diff
@@ -1618,7 +1622,15 @@ public void nationSet(Player player, String[] split) throws TownyException, Inva
 
-			} else if (split[0].equalsIgnoreCase("taxes")) {
+			} else if (split[0].equalsIgnoreCase("spawn")){
+				try{
```
This will require a permission node check here
```
if (!TownyUniverse.getPermissionSource().testPermission(player, PermissionNodes.TOWNY_COMMAND_NATION_SET_SPAWN.getNode()))
 						throw new TownyException(TownySettings.getLangString("msg_err_command_disable"));```

LlmDl commented on src/com/palmergames/bukkit/towny/command/NationCommand.java:
```diff
@@ -1847,4 +1859,211 @@ public void nationRename(Player player, Nation nation, String newName) {
+
+                nationSpawn(player, split, nation, notAffordMSG);
+
+            } else {
```
I'm not certain that this is something needed. This section of town spawning is for public spawning, which I haven't seen yet if you've set up public/non-public settings for nations. I don't think that the intent of the tickets requesting this was for spawning to other players' nations, just to a player's own nation.

LlmDl commented:
Didn't find very much to complain about on this read-through. It looks like removing the handling for /nation spawn {othernation} is needed.

Other than that, it appears that nation spawns can be set in any town inside of the nation? I'd always pictured it being something that could only be set in the capital city.

MartenM commented:
> Didn't find very much to complain about on this read-through. It looks like removing the handling for /nation spawn {othernation} is needed.
Uhm, why would this have to be removed? I imanged it would be pretty cool if you could visit the spawns of other nations.

> Other than that, it appears that nation spawns can be set in any town inside of the nation? I'd always pictured it being something that could only be set in the capital city
Oh yeah, that makes sense. Would make capital towns a bit more important as well.
Will implement that someday
.

LlmDl commented:
If you want to implement the 2nd thing with a config option under the global nation settings section that would work too.

As for the public spawning thing, you'd have to add isPublic to nations in the database and in the /nation status screen, and also an /n toggle public command.

MartenM commented:
> If you want to implement the 2nd thing with a config option under the global nation settings section that would work too.

Sure, that sounds fine. Will look into it.

> As for the public spawning thing, you'd have to add isPublic to nations in the database and in the /nation status screen, and also an /n toggle public command.

Might as well do that then. This could also be used by feature additions.

MartenM commented:
**Summary:**
1. Added nation isPublic var.
2. Added the nation isPublic var to the databases.
3. Added the ability to toggle the nation public setting.
4. Added a check that makes sure you cannot teleport to non-public nations.
5. Added permission nodes.
6. Added messages.

**Needs testing:**
3

I think that was all of it?
If you could check the code, that would be awesome :wink:

LlmDl commented:
First read-over of the new stuff was good, let me know how the testing goes.

LlmDl commented:
Fairly sure my IDE has tabs equaling out to 4 spaces, but there's been enough people working on things (and people with intellij auto-formatting things,) that there's some inconsistency now.

MartenM commented:
(and people with intellij auto-formatting things) well yeah, that's also me, but I try to make sure it replaces it all after I use tabs.

As far as the pull-request I think it's somewhat finished. If you could take a look at it that would be nice.
After that I will add the language lines to the other language files as well and then we can merge this with the dev branch I think.

MartenM commented:
I think that this can be considered done for right now.

LlmDl commented:
This will be accepted soon, just can't do it right away.
Note to self: -add new nodes to proper locations in plugin.yml

MartenM commented:
soon™

Sorry, to bug you, but any chance that this is getting pulled any time soon?

LlmDl commented:
Sorry this took as long as it did, we managed to get it all sorted right before work got very busy for me. This is the busiest time of the year for me. I put out 0.92.0.11 earlier today and this will be a part of 0.92.0.12 which I hope to also release today.

MartenM commented:
> Sorry this took as long as it did, we managed to get it all sorted right before work got very busy for me. This is the busiest time of the year for me. I put out 0.92.0.11 earlier today and this will be a part of 0.92.0.12 which I hope to also release today.

No worries. Thank you for the merge :D
